### PR TITLE
1️⃣ refactor(services: rate-limit): migrate from fastify to @whatwg-node/server

### PR DESCRIPTION
**Pull request by YassinEldeeb**
related to https://github.com/kamilkisiela/graphql-hive/issues/657 but doesn't close it entirely

### CODE REVIEW

File: packages/services/rate-limit/src/index.ts [hunk 1]
```diff
@@ -24,67 +23,28 @@ async function main() {
...
-      rateLimitConfig: {
-        interval: env.limitCacheUpdateIntervalMs,
-      },
-      rateEstimator: env.hiveServices.usageEstimator,
-      emails: env.hiveServices.emails ?? undefined,
-      storage: {
-        connectionString: createConnectionString(env.postgres),
-      },
-    });
+  const logger = createLogger() as FastifyLoggerInstance;
```
Comment (dotansimha) [inline packages/services/rate-limit/src/index.ts:26]: If you are dropping the `fastify` dependency, I think you should also get rid of using the `FastifyXYZ` types?

File: packages/services/rate-limit/src/index.ts [hunk 2]
```diff
@@ -24,67 +23,28 @@ async function main() {
...
+  try {
     registerShutdown({
-      logger: server.log,
+      logger,
       async onShutdown() {
-        await Promise.all([ctx.stop(), server.close()]);
-      },
-    });
-
-    server.route({
```
Comment (dotansimha) [inline packages/services/rate-limit/src/index.ts:63]: Why did you remove these? they are used by the K8s and Docker for health and readiness

File: packages/services/service-common/src/index.ts [hunk 3]
```diff
@@ -5,3 +5,4 @@ export * from './metrics';
 export * from './heartbeats';
```
Comment (dotansimha) [inline packages/services/service-common/src/index.ts:5]: I assume we should remove `FastifyLoggerInstance` ?

File: packages/services/rate-limit/package.json [hunk 4]
```diff
@@ -15,10 +15,13 @@
     "@sentry/tracing": "7.21.1",
     "@trpc/client": "10.4.3",
     "@trpc/server": "10.4.3",
+    "@valu/trpc-fetch-api-adapter": "^0.0.3",
     "@whatwg-node/fetch": "0.5.3",
+    "@whatwg-node/server": "^0.4.17",
     "date-fns": "2.29.3",
     "dotenv": "10.0.0",
     "got": "12.5.3",
+    "itty-router": "2.6.6",
```
Comment (dotansimha) [inline packages/services/rate-limit/package.json:24]: @ardatan you mentioned that `itty-router` have some issues recently? should we use it here?

**Review by dotansimha — CHANGES_REQUESTED** [review]
Please refer to the comments

File: packages/services/rate-limit/src/index.ts (hunk 1 above)
Comment (YassinEldeeb) [reply packages/services/rate-limit/src/index.ts:26]: yes, the only concern would be to ensure that all services have been migrated to `whatwg` because I think there maybe is 1 or 2 that are still using fastify which would cause a type error when we change the arg type of `logger` for `registerShutdown` in `services/service-common`

File: packages/services/rate-limit/src/index.ts (hunk 2 above)
Comment (YassinEldeeb) [reply packages/services/rate-limit/src/index.ts:63]: they haven't been removed but extracted to an external router for simplifying the testing process in the future by using the router without the need to launch the server.

https://github.com/kamilkisiela/graphql-hive/pull/858/files#diff-783f53b46dea4ebd59badd5a393cefdb1b6bca12a68e67324e6ffdcbbf89473eR34-R55

File: packages/services/service-common/src/index.ts (hunk 3 above)
Comment (YassinEldeeb) [reply packages/services/service-common/src/index.ts:5]: this's the type we're using for the `registerShutdown` `logger` arg to avoid erroring for type conflicts, as you've mentioned above, there's going to be no need for it when we ensure all services have been migrated so that we can safely change the type for it knowing every service that's using it is passing the right arg type

File: packages/services/rate-limit/package.json (hunk 4 above)
Comment (YassinEldeeb) [reply packages/services/rate-limit/package.json:24]: curious about that 👀

File: packages/services/rate-limit/src/index.ts (hunk 1 above)
Comment (dotansimha) [reply packages/services/rate-limit/src/index.ts:26]: I see. I'm also fine with grouping all these PRs into one if they are all similar and does the same thing - and then we can just drop `fastify` types in the same PR?

File: packages/services/rate-limit/src/index.ts (hunk 2 above)
Comment (dotansimha) [reply packages/services/rate-limit/src/index.ts:63]: ok cool!

File: packages/services/service-common/src/index.ts (hunk 3 above)
Comment (dotansimha) [reply packages/services/service-common/src/index.ts:5]: Got it, see my comment above.

File: packages/services/rate-limit/src/index.ts (hunk 1 above)
Comment (YassinEldeeb) [reply packages/services/rate-limit/src/index.ts:26]: I was planning to open a single PR with everything but @n1ru4l advised to open a PR per service.

image

**Comment by YassinEldeeb** [comment]
:one:

**Comment by YassinEldeeb** [comment]
I accidentally, discarded the changes in this branch while I was syncing my fork with the latest of the original repo, I'm gonna open another PR with the same content.
